File: DSA/done/1291 - sequentialDigits.cpp
// https://leetcode.com/problems/sequential-digits/

#include <vector>
#include <string>
#include <algorithm>

class Solution {
public:
    std::vector<int> sequentialDigits(int low, int high) {
        int min_len = std::to_string(low).length();
        // Clamp max_len to 9 because "123456789" only has 9 characters
        int max_len = std::min(9, (int)std::to_string(high).length());
        
        std::string mas_str = "123456789";
        std::vector<int> res;
        
        // Outer loop for the window sizes
        for (int i = min_len; i <= max_len; ++i) {
            int n = mas_str.length();
            
            // Cast to int to prevent unsigned underflow bugs
            for (int l = 0; l <= n - i; ++l) {
                std::string sub = mas_str.substr(l, i);
                int seq_dig = std::stoi(sub);
                
                if (seq_dig >= low && seq_dig <= high) {
                    res.push_back(seq_dig);
                }
                
                if (seq_dig > high) {
                    break;
                }
            }
        }
        
        return res;
    }
};

File: DSA/done/1291 - sequentialDigits.py
# https://leetcode.com/problems/sequential-digits

from typing import List

class Solution:
    def sequentialDigits(self, low: int, high: int) -> List[int]:
        min_len, max_len = len(str(low)), len(str(high))
        mas_str = "123456789"
        res = []
        
        for i in range(min_len, max_len + 1):
            l = 0
            # len(mas_str) + 1 ensures the right pointer 'r' can reach index 9 for slicing
            for r in range(i, len(mas_str) + 1):
                seq_dig = int(mas_str[l:r])
                
                if low <= seq_dig <= high: 
                    res.append(seq_dig)
                
                # Optional Optimization: if it's already too big, stop checking this length
                if seq_dig > high:
                    break
                    
                l += 1
        
        return res